### PR TITLE
blob/all: add support for ContentMD5 integrity checks

### LINKED ISSUE #569: blob: add support for content-md5
### Is your feature request related to a problem? Please describe.

add support for [Content-md5](https://tools.ietf.org/html/rfc1864)

### Describe alternatives you've considered

```go
type WriterOptions struct{
    ContentMD5 string
    ...
}
```

### PR DESCRIPTION
**Pull request by vangent**
Fixes #569.

This Pull Request is a clone of #576 by @myml.

I've made a few minor changes, plus merged from master and recorded golden files for the GCS and S3 tests.

### CODE REVIEW

**Comment by vangent** [comment]
@myml, can you review my changes? I made minor changes and added the golden files for GCS and S3 tests that you couldn't add. Sorry, I didn't have permissions to modify your Pull Request directly to add the golden files. Thanks for the contribution!

Please also confirm that you are OK with your commits being contributed to this project per the googlebot comment a minute ago.

**Comment by myml** [comment]
It is ok,thank you for your help.

> @myml，你能看看我的变化吗？我做了一些小改动，并添加了无法添加的GCS和S3测试的黄金文件。抱歉，我没有权限直接修改您的Pull请求以添加黄金文件。谢谢你的贡献！
> 
> 请确认您可以在一分钟前通过googlebot评论为此项目提交您的提交。

**Comment by vangent** [comment]
Setting CLA to yes because @myml is OK with this contribution.

File: blob/fileblob/fileblob.go [hunk 1]
```diff
@@ -250,22 +254,34 @@ func (b *bucket) NewTypedWriter(ctx context.Context, key string, contentType str
...
+	path       string
+	attrs      xattrs
+	contentMD5 []byte
+	md5hash    hash.Hash
 }
 
 func (w writer) Write(p []byte) (n int, err error) {
+	if w.md5hash != nil {
+		if _, err := w.md5hash.Write(p); err != nil {
+			return 0, err
```
Comment (ijt) [inline blob/fileblob/fileblob.go]: The returned err could probably use some context.

`return 0, fmt.Errorf("updating md5 hash: %v", err)`

Comment (vangent) [reply blob/fileblob/fileblob.go]: Done.

**Pull request merged**